File: pa2/backup/Lex.c
//Lex.c
//This program acts like the client for List.c and sorts the input file in lexographical order by line.
//Written by Samuel Shin, sayshin@ucsc.edu, pa2.

#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include"List.h"

#define MAX_LEN 160
void sort(char* x, List y);

int main (int argc, char * argv[])
{
	FILE *in, *out;
	char line[MAX_LEN];
	int linenum;
	int temp = 0;
	int j = 0; 
	int pass, result;
	
	if (argc != 3)
	{
		printf("Usage: %s <input file> <output file> \n", argv[1]);
		exit(1);
	}

	in = fopen(argv[1], "r");
	out = fopen(argv[2], "w");
	if (in == NULL)
	{
		printf("Unable to open file %s for reading \n", argv[1]);
		exit(1);
	}

	if (out == NULL)
        {
                printf("Unable to open file %s for reading \n", argv[2]);
                exit(1);
        }
	
	while (fgets(line, sizeof(line), in))
	{
		linenum++;
	}
	
	rewind(in);
	fclose(in);
	char* arr[linenum];
	int x = 0;
	in  = fopen(argv[1], "r");
	
	if (in == NULL)
        {
                printf("Unable to open file %s for reading \n", argv[1]);
                exit(1);
        }

	while (fgets(line, sizeof(line), in) != NULL)
	{
		arr[x] = calloc(strlen(line) + 1, 1);
		strcpy(arr[x], line);
		x++;
	}
       
	List list = newList();	
	//sort(*arr, list);
	int len =  (sizeof(arr) / sizeof(arr[0]));
	for (int i = 1; i < len; i++)
	{
		temp = i;
		j = i -1;
		moveBack(list);
		pass = 0;
		
		if (length(list) == 0)
		{
			result = strcmp(arr[j], arr[temp]);
			if (result <= 0)
			{
				prepend(list, j);
				append(list, temp);
				continue;
			}
			else 
			{
				prepend(list, temp);
				append(list, j);
				continue;
			}
		}
		while (j >= 0)
		{
			result = strcmp(arr[temp], arr[j]);
			if (result < 0)
			{
				movePrev(list);
			}
			if (index(list) == -1)
			{
				prepend(list, temp);
				pass = 1;
			}
			j--;
		}
		if (pass == 0)
		{
			insertAfter(list, temp);
		}
	}

	int a = 0;
	moveFront(list);
	while (index(list) >= 0)
	{
		a = get(list);
		fprintf(out, "%s", arr[a]);
		//fprintf(out, "\n");
		moveNext(list);
	}
	free(arr);
	freeList(&list);
	fclose(in);
	fclose(out);
	return (0);
}

/*void sort (char*A, List L)
{
	int temp = 0;
	int j = 0;
	int pass, result;
	//int len =  sizeof(A);
	//printf("%d", len);
	for (int i = 1; i < 12; i++)
	{
		temp = i;
		j = i - 1;
		moveBack(L);
		pass = 0;
		
		if (length(L) == 0)
		{
			char arrj[sizeof(A[j])];
                	char arrtemp[sizeof(A[temp])];
                	char strj[sizeof(A[j])];
                	char strtemp[sizeof(A[temp])];
                	*strj = A[j];
                	*strtemp = A[temp];
                	strcpy(arrj, strj);
                	strcpy(arrtemp, strtemp);

                	result = strcmp(arrj, arrtemp);
			if (result < 0)
			{
				prepend(L, j);
				append(L, temp);
				continue;
			}
			else 
			{
				prepend(L, temp);
				append(L, j);
				continue;
			}
		}
		while (j >= 0)
                {
			char arrj[sizeof(A[j])];
                	char arrtemp[sizeof(A[temp])];
                	char strj[sizeof(A[j])];
                	char strtemp[sizeof(A[temp])];
                	*strj = A[j];
                	*strtemp = A[temp];
                	strcpy(arrj, strj);
                	strcpy(arrtemp, strtemp);
			
                	result = strcmp(arrtemp, arrj);
                        if (result < 0)
                        {
                                movePrev(L);
                        }
                        if (index(L) == -1)
                        {
                                prepend(L, temp);
                                pass = 1;
                        }
                        j--;
                }
                if (pass == 0)
                {
                        insertAfter(L, temp);
                }	
	}
}*/

File: pa1/backup/Lex.java
//Lex.java
//This program reads in an input file and produces and output file 
//that sorts the input file in alphabetical order by line by using 
//the List ADT that is defined in List.java
//Samuel Shin, sayshin@ucsc.edu, PA1
import java.io.*;
import java.util.*;

class Lex
{
	public static void main (String[] args) throws IOException
	{
		Scanner in = null;
		PrintWriter out = null;
		String line = null;
		int linenum = 0;		
		int i = 0;

		if (args.length < 2)
		{
			System.err.println("Usage: FileIO infile outfile");
			System.exit(1);
		}

		in = new Scanner(new File(args[0]));
		out = new PrintWriter(new FileWriter(args[1]));

		while (in.hasNextLine())
		{
			linenum++;
			in.nextLine();
		}
		
		String[] arr = new String[linenum];
		in = new Scanner(new File(args[0])); //reset the scanner to look at top of file again
		
		while (in.hasNextLine())
		{
			line = in.nextLine();
			arr[i] = line;
			i++;				
		}
		
		List list = new List();

		sort(arr, list);
		int a = 0;
                list.moveFront();
                while (list.index() >= 0)
                {
                        a = list.get();
                        out.println(arr[a]);
			out.flush();
                        list.moveNext();
                }
		
		in.close();
		out.close();
		
	}	

	static void sort (String[] A, List L)
	{
		int temp = 0;
		int j = 0;
		int temp2 = 0;
		boolean pass;
		for (int i = 1; i < A.length; i++)
		{
			temp = i;
			j = i -1;
			L.moveBack();
			pass = false;
			if (L.length() == 0)
                        {
                                if (A[j].compareTo(A[temp]) < 0)
                                {
                                        L.prepend(j);
                                        L.append(temp);
					continue;
                                }
                                else
                                {
                                        L.prepend(temp);
                                        L.append(j);
					continue;
                                }
                        }
			while (j >= 0)
			{
				if (A[temp].compareTo(A[j]) < 0)
				{
					L.movePrev();
				}
				if (L.index() == -1)
				{
					L.prepend(temp);
					pass = true;
				}
				j--;	
			}
			if (pass == false)
			{
				L.insertAfter(temp);
			}
		}
	}

}
